[stdlib] SIMD conformance to EqualityComparable

helehex opened the pull request:
This allows `SIMD` to conform to `EqualityComparable`, without losing any of the original behavior.

It uses the 4th overload resolution rule to give the new methods lower precedence, while still conforming to `EqualityComparable`

gabrieldemarmiesse reviewed (commented):
I wouldn't be against some more details (in the docs or something) since this seems a bit like black magic and using return types for overload resolution :)
Maybe the docstring would be a good place to explain how to actually use this overload?

JoeLoser reviewed (commented):
This is a pretty neat way to make `SIMD` conform to `EqualityComparable`, props to your cleverness!

This seems like something we should do in the short-term until we have conditional conformance and other things.  Accepting this PR would unlock a bunch of other future nice things.  To start with, we could then push this `EqualityComparable` requirement into `CollectionElement`, which would then mean we can now implement `__eq__ ` and ` __ne__` for collections.  Further, it would allow us to simplify things like `List.count` and `List.index` which currently use an ad-hoc composite trait to ensure the given type implements `__eq__` since that isn’t a requirement of `CollectionElement`.

I've added this PR to bring up with the team on Monday during our internal design discussions, and I'll circle back after we meet.  But this is super cool, thanks for pushing on this!

JoeLoser reviewed (commented):
!sync

helehex committed:
simd conformance to EqualityComparable

Signed-off-by: Max Brylski <helehex@gmail.com>

helehex committed:
Removed `String` overloads for `assert_equal()`
Swapped test arguments around so that the literal is always on the right. This was a way to have the testable type be inferable.

Signed-off-by: Max Brylski <helehex@gmail.com>
